server/sonar-server/src/main/java/org/sonar/server/projectanalysis/ws/SearchAction.java

```diff
@@ -0,0 +1,156 @@
+import static org.sonar.core.util.stream.Collectors.toOneElement;
+import static org.sonar.db.component.SnapshotQuery.SORT_FIELD.BY_DATE;
+import static org.sonar.db.component.SnapshotQuery.SORT_ORDER.DESC;
+import static org.sonar.server.ws.WsUtils.writeProtobuf;
+import static org.sonarqube.ws.client.projectanalysis.EventCategory.OTHER;
+import static org.sonarqube.ws.client.projectanalysis.ProjectAnalysesWsParameters.PARAM_CATEGORY;
+import static org.sonarqube.ws.client.projectanalysis.ProjectAnalysesWsParameters.PARAM_PROJECT;
+import static org.sonarqube.ws.client.projectanalysis.SearchRequest.DEFAULT_PAGE_SIZE;
+
+public class SearchAction implements ProjectAnalysesWsAction {
+  private final DbClient dbClient;
+  private final ComponentFinder componentFinder;
+  private final UserSession userSession;
+
+  public SearchAction(DbClient dbClient, ComponentFinder componentFinder, UserSession userSession) {
+    this.dbClient = dbClient;
+    this.componentFinder = componentFinder;
+    this.userSession = userSession;
+  }
+
+  @Override
+  public void define(WebService.NewController context) {
+    WebService.NewAction action = context.createAction("search")
+      .setDescription("Search a project analyses and attached events.<br>" +
+        "Requires the following permission:" +
+        "<ul>" +
+        "  <li>'Browse' on the specified project</li>" +
+        "</ul>")
+      .setSince("6.3")
+      .setResponseExample(getClass().getResource("search-example.json"))
+      .setHandler(this);
+
+    action.addPagingParams(DEFAULT_PAGE_SIZE, 500);
+
+    action.createParam(PARAM_PROJECT)
+      .setDescription("Project key")
+      .setRequired(true)
+      .setExampleValue(KeyExamples.KEY_PROJECT_EXAMPLE_001);
+
+    action.createParam(PARAM_CATEGORY)
+      .setPossibleValues(EnumSet.allOf(EventCategory.class))
+      .setExampleValue(OTHER.name());
+  }
+
+  @Override
+  public void handle(Request request, Response response) throws Exception {
+    ProjectAnalyses.SearchResponse searchResponse = Stream.of(request)
+      .map(toWsRequest())
+      .map(this::search)
+      .map(new SearchResponseBuilder().buildWsResponse())
+      .collect(toOneElement());
+    writeProtobuf(searchResponse, request, response);
+  }
+
+  private static Function<Request, SearchRequest> toWsRequest() {
+    return request -> {
+      String category = request.param(PARAM_CATEGORY);
+      return SearchRequest.builder()
+        .setProject(request.mandatoryParam(PARAM_PROJECT))
+        .setCategory(category == null ? null : EventCategory.valueOf(category))
+        .setPage(request.mandatoryParamAsInt(Param.PAGE))
+        .setPageSize(request.mandatoryParamAsInt(Param.PAGE_SIZE))
+        .build();
+    };
+  }
+
+  private SearchResults search(SearchRequest request) {
+    try (DbSession dbSession = dbClient.openSession(false)) {
+      return Stream.of(SearchResults.builder(dbSession, request))
+        .peek(addProject(dbSession, request))
```

Comment:
similar to what I have done with the response, I understand. I thought about it, will give it another try